Implement NeuralStatistician

masenka31 opened the pull request:
Adding a new generative model working on sets of data. Work in progress now.

masenka31 committed:
neural statistician

masenka31 committed:
polished

nmheim commented on src/statistician.jl:
```diff
@@ -0,0 +1,101 @@
+- β2: KL[q(z|c,x) || p(z|c)]
+"""
+
+function lossNS(x, model::Statistician;β1=1.0,β2=1.0)
```
It would be nice to have (a short version) of the loss in the docs. I have not read the paper but it is probably the ELBO that you are optimizing here? In that case it would be nice to just rename the function to `elbo`.

All other models have their loss defined as `elbo(m::ModelName, x::AbstractArray)`, so please switch the arugments here as well.

masenka31 replied:
It is ELBO, I changed it. Still, it has to be long, because the ELBO evaluation has more parts then simple VAE.

nmheim commented on src/statistician.jl:
```diff
@@ -0,0 +1,101 @@
+    z = rand(model.encoder_z, h)
+	
+    # 3 loss terms - likelihood, kl1, kl2
+	llh = mean(logpdf(model.decoder, x, z))
```
Align all lines to 4 spaces indentation here (and also in the rest of the code)

nmheim commented on src/statistician.jl:
```diff
@@ -0,0 +1,101 @@
+"""
+    Statistician()
```
Rename to `NeuralStatistician` and add a reference to the paper

nmheim reviewed (changes requested):
Good stuff, thanks a lot for the PR! I requested some minor changes in the code :)

masenka31 committed:
statistician update, elbo, docs, test

masenka31 committed:
indentation

nmheim commented on src/statistician.jl:
```diff
@@ -12,90 +14,87 @@ There are six parts:
 function as
     
-    `loss(x) = lossNS(x, NS)`
+    `loss(x) = -elbo(NeuralStatistician,x)`
```
This could be misunderstood. It looks like the `elbo` would only get a `NeuralStatistician` type instead of an actual object. maybe something like this:
```julia
model = NeuralStatistician(cdim, instance_enc, enc_c_dist, cond_z_dist, enc_z_dist, dec_dist)
loss(x) = - elbo(model,x)
```
In that case it would also be really great to have short explanations of the variables in there like `instance_enc`. but if you are pressed with time we can just merge it with the corrections above.

nmheim reviewed (commented):
Great, we are almost there now! In order to create a new release you just need to increase the version number of GenerativeModels in the Project.toml (from 0.2.2 to 0.2.3). then the julia tagbot will automatically create a new release an register it in the julia registry, so that we can install/update the package via `]add Generativemodels`/`]update` :)

masenka31 committed:
NS docs update, swap IE and pc positions, version update

masenka31 commented:
I updated the documentation for NeuralStatistician to include an example and rewrote the docs for `elbo` to be more accurate. I think there should be no more problems.

masenka31 committed:
test error resolved

nmheim commented:
Looks great, thank you! :)
